pkg/controllers/infracluster/infracluster_controller.go

```diff
@@ -0,0 +1,288 @@
+package infracluster
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	"github.com/go-logr/logr"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"k8s.io/client-go/rest"
+	awsv1 "sigs.k8s.io/cluster-api-provider-aws/v2/api/v1beta2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/cluster-capi-operator/pkg/controllers"
+	"github.com/openshift/cluster-capi-operator/pkg/operatorstatus"
+)
+
+const (
+	// Controller conditions for the Cluster Operator resource
+	InfraClusterControllerAvailableCondition = "InfraClusterControllerAvailable"
+	InfraClusterControllerDegradedCondition  = "InfraClusterControllerDegraded"
+
+	defaultCAPINamespace    = "openshift-cluster-api"
+	defaultInfraClusterName = "cluster"
+	clusterOperatorName     = "cluster-api"
+	// This is the managedByAnnotation value that this controller sets by default when it creates an InfraCluster object.
+	// If the managedByAnnotation key is set, and it has this as the value, it means this controller is managing the InfraCluster.
+	managedByAnnotationValueClusterCAPIOperator = ""
+)
+
+type InfraClusterController struct {
+	operatorstatus.ClusterOperatorStatusClient
+	Scheme       *runtime.Scheme
+	Images       map[string]string
+	RestCfg      *rest.Config
+	Platform     configv1.PlatformType
+	Infra        configv1.Infrastructure
+	InfraCluster client.Object
+}
+
+// Reconcile reconciles the cluster-api ClusterOperator object.
+func (r *InfraClusterController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx).WithName("InfraClusterController")
+
+	res, err := r.reconcile(ctx, log)
```

Comment:
Yeah, this is not following that convention because we don't actually have a concrete type to get, but rather before knowing what object to get we need to switch over the platform and then kick off the specific reconciliation function for the concrete InfraCluster type (e.g. `reconcileAWSCluster()`).
So we came up with this which is the cleanest way of reconciling this that we could think at the time.